ttnn/CMakeLists.txt

```diff
@@ -478,17 +519,71 @@ if(WITH_PYTHON_BINDINGS)
     list(APPEND TTNN_PUBLIC_LINK_LIBRARIES pch_pybinds ${Python3_LIBRARIES})
 endif()
 
-add_library(ttnn SHARED ${TTNN_FINAL_SRC})
+message (STATUS "CMAKE COMPILER ID ${CMAKE_CXX_COMPILER_ID}")
+
+function(use_precompiled_header TARGET)
+    message(STATUS "PRECOMPILED_HEADER_TARGET: ${PRECOMPILED_HEADER_TARGET}")
+    if("${PRECOMPILED_HEADER_TARGET}" STREQUAL "")
+        set(PRECOMPILED_HEADER_TARGET "${TARGET}" PARENT_SCOPE)
+        target_precompile_headers(${TARGET} PUBLIC ${TTNN_PRECOMPILED_HEADERS})
+    else()
+        target_precompile_headers(${TARGET} REUSE_FROM ${PRECOMPILED_HEADER_TARGET})
+    endif()
+endfunction()
+
+set(PRECOMPILED_HEADER_TARGET "")
+
+function(add_ttnn_sublibrary SUBLIBRARY_NAME)
+    add_library(${SUBLIBRARY_NAME} SHARED ${ARGN})
+    target_include_directories(${SUBLIBRARY_NAME} PRIVATE ${TTNN_PUBLIC_INCLUDE_DIRS})
+    target_link_libraries(${SUBLIBRARY_NAME} PRIVATE ${TTNN_PUBLIC_LINK_LIBRARIES})
+    target_link_directories(${SUBLIBRARY_NAME} PUBLIC ${TTNN_PUBLIC_LINK_DIRS})
+    use_precompiled_header(${SUBLIBRARY_NAME})
+    set(PRECOMPILED_HEADER_TARGET "${PRECOMPILED_HEADER_TARGET}" PARENT_SCOPE)
+    target_compile_options(${SUBLIBRARY_NAME} PRIVATE -MP -Wno-int-to-pointer-cast -fno-var-tracking -ffunction-sections)
+    set_target_properties(${SUBLIBRARY_NAME} PROPERTIES  DEFINE_SYMBOL "")
+    if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+        target_compile_definitions(${SUBLIBRARY_NAME} PRIVATE DISABLE_NAMESPACE_STATIC_ASSERT)
+        target_link_options(${SUBLIBRARY_NAME} PRIVATE "-Wl,--gc-sections" "-Wl,--no-as-needed")
+    endif()
+    install(TARGETS ${SUBLIBRARY_NAME}
+    LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+    ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+    COMPONENT tt_build_artifacts
+    )
+endfunction()
+
+foreach(SUBLIBRARY_PATH ${TTNN_SUBLIBRARIES})
+    set(SUBLIBRARY_SRCS ${TTNN_SRC})
+    list(FILTER SUBLIBRARY_SRCS INCLUDE REGEX "${SUBLIBRARY_PATH}")
+    list(FILTER TTNN_SRC EXCLUDE REGEX "${SUBLIBRARY_PATH}")
+    if(SUBLIBRARY_SRCS)
+        string(REPLACE "/"  "_"  SUBLIBRARY_NAME ${SUBLIBRARY_PATH})
+        add_ttnn_sublibrary(${SUBLIBRARY_NAME} ${SUBLIBRARY_SRCS})
+        list(APPEND ALL_TTNN_SUBLIBRARIES ${SUBLIBRARY_NAME})
+    endif()
+endforeach(SUBLIBRARY_PATH ${TTNN_SUBLIBRARIES})
+
+add_ttnn_sublibrary(ttnn_tensor ${TENSOR_SRCS})
+add_ttnn_sublibrary(tt_dnn ${TT_DNN_SRCS})
+
+add_library(ttnn SHARED ${TTNN_BASE_SRCS} ${QUEUE_SRCS})
 target_compile_options(ttnn PUBLIC -MP -Wno-int-to-pointer-cast -fno-var-tracking)
 
 if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
     target_compile_definitions(ttnn PUBLIC DISABLE_NAMESPACE_STATIC_ASSERT)
+    target_link_options(ttnn PUBLIC  "-Wl,--gc-sections" "-Wl,--no-as-needed")
```

Comment:
Why are you adding these flags?

Reply:
The sub libraries include header files from other sub libraries, leading to undefined symbols in the final shared object. When the tests executables are compiled, all the shared libraries are added as inputs. 

GCC picks up the symbols in the order that the inputs are passed to the command. This leads to undefined symbols  in the executable, leading to a linker error. 

These flags remove the undefined symbols from the shared object library, preventing this error. 

Clang doesn't have this problem.

Reply:
These flags don't remove the undefined symbols, they prevent the linker from excluding the libraries that are linked but not used. I assume that's because your PR created artificial sublibraries that don't properly link each other. So you broke the build process in subtle ways that made adding the `--no-as-needed` flag necessary. The correct fix is to track which sublibrary needs to link which other sublibrary to ensure no undefined symbols are present.

Reply:
Do you mean that when each sub library is made, it has no undefined symbols? 

As each sub library corresponds to particular op group, and many ops call other ops, this could become complex. Additionally, I think it would lead to multiple definition errors if the same source file is included in both sub libraries. 

I would not want to make the build process dependent on the code structure.